### ISSUE TITLE
Exception occurred: System.ObjectDisposedException: Cannot access a disposed object.

### ISSUE THREAD

**Issue by D3m0nKingx** [stratisproject/StratisBitcoinFullNode#797]
After terminating the Node (Ctrl+c), the following exception occurred:

```bash
fail: Stratis.Bitcoin.Features.MemoryPool.MempoolBehavior[0]
      Exception occurred: System.ObjectDisposedException: Cannot access a disposed object.
      Object name: 'LoggerFactory'.
         at Microsoft.Extensions.Logging.LoggerFactory.CreateLogger(String categoryName)
         at Stratis.Bitcoin.Features.Consensus.ConsensusSettings..ctor(NodeSettings nodeSettings, ILoggerFactory loggerFactory) in /home/ader/coins/StratisBitcoinFullNode/src/Stratis.Bitcoin.Features.Consensus/ConsensusSettings.cs:line 36
      --- End of stack trace from previous location where exception was thrown ---
         at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
         at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProvider provider)
         at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(IServiceCallSite callSite, TArgument argument)
         at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
         at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitSingleton(SingletonCallSite singletonCallSite, ServiceProvider provider)
         at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(IServiceCallSite callSite, TArgument argument)
         at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProvider provider)
         at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(IServiceCallSite callSite, TArgument argument)
         at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
         at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitSingleton(SingletonCallSite singletonCallSite, ServiceProvider provider)
         at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(IServiceCallSite callSite, TArgument argument)
         at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProvider provider)
         at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(IServiceCallSite callSite, TArgument argument)
         at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
         at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitSingleton(SingletonCallSite singletonCallSite, ServiceProvider provider)
         at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(IServiceCallSite callSite, TArgument argument)
         at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteExpressionBuilder.<>c__DisplayClass14_0.<Build>b__0(ServiceProvider provider)
         at Microsoft.Extensions.DependencyInjection.ServiceProvider.GetService(Type serviceType)
         at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetService[T](IServiceProvider provider)
         at Stratis.Bitcoin.FullNode.NodeService[T](Boolean failWithDefault) in /home/ader/coins/StratisBitcoinFullNode/src/Stratis.Bitcoin/FullNode.cs:line 91
         at Stratis.Bitcoin.Base.ChainState.get_IsInitialBlockDownload() in /home/ader/coins/StratisBitcoinFullNode/src/Stratis.Bitcoin/Base/ChainState.cs:line 84
         at Stratis.Bitcoin.Features.MemoryPool.MempoolBehavior.<ProcessInvAsync>d__29.MoveNext() in /home/ader/coins/StratisBitcoinFullNode/src/Stratis.Bitcoin.Features.MemoryPool/MempoolBehavior.cs:line 346
      --- End of stack trace from previous location where exception was thrown ---
         at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
         at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
         at Stratis.Bitcoin.Features.MemoryPool.MempoolBehavior.<AttachedNode_MessageReceivedAsync>d__26.MoveNext() in /home/ader/coins/StratisBitcoinFullNode/src/Stratis.Bitcoin.Features.MemoryPool/MempoolBehavior.cs:line 193

Unhandled Exception: System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'LoggerFactory'.
   at Microsoft.Extensions.Logging.LoggerFactory.CreateLogger(String categoryName)
   at Stratis.Bitcoin.Features.Consensus.ConsensusSettings..ctor(NodeSettings nodeSettings, ILoggerFactory loggerFactory) in /home/ader/coins/StratisBitcoinFullNode/src/Stratis.Bitcoin.Features.Consensus/ConsensusSettings.cs:line 36
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(IServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitSingleton(SingletonCallSite singletonCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(IServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(IServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitSingleton(SingletonCallSite singletonCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(IServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(IServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitSingleton(SingletonCallSite singletonCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(IServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteExpressionBuilder.<>c__DisplayClass14_0.<Build>b__0(ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.GetService(Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetService[T](IServiceProvider provider)
   at Stratis.Bitcoin.FullNode.NodeService[T](Boolean failWithDefault) in /home/ader/coins/StratisBitcoinFullNode/src/Stratis.Bitcoin/FullNode.cs:line 91
   at Stratis.Bitcoin.Base.ChainState.get_IsInitialBlockDownload() in /home/ader/coins/StratisBitcoinFullNode/src/Stratis.Bitcoin/Base/ChainState.cs:line 84
   at Stratis.Bitcoin.Features.MemoryPool.MempoolBehavior.<ProcessInvAsync>d__29.MoveNext() in /home/ader/coins/StratisBitcoinFullNode/src/Stratis.Bitcoin.Features.MemoryPool/MempoolBehavior.cs:line 346
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Stratis.Bitcoin.Features.MemoryPool.MempoolBehavior.<AttachedNode_MessageReceivedAsync>d__26.MoveNext() in /home/ader/coins/StratisBitcoinFullNode/src/Stratis.Bitcoin.Features.MemoryPool/MempoolBehavior.cs:line 209
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
```

**Comment by Aprogiena** [comment]
Thank you, I can confirm that I have received this exact exception as well.

**Comment by Aprogiena** [comment]
So what we've got here is that MemorePool.ProcessInvAsync is executed, which calls `if (this.chainState.IsInitialBlockDownload)`, which fails because the components are disposed already. This could very much be related to #679 because a peer obviously sent us `inv` message after we triggered shutdown. So I suppose that once that issue with ConnectionManager is solved, no longer we will be accepting new messages from our peers at this moment, so this issue could be automatically fixed by fixing #679 without any additional effort.

So let's keep an eye on #679 and once it is closed, let's see if we get this one again.

**Comment by fassadlr** [comment]
I have observed this issue whilst running the node as well.

**Comment by fassadlr** [comment]
@Aprogiena @noescape00 @noescape00 guys Im pretty sure that #825 fix this bug?

**Comment by Aprogiena** [comment]
i have no idea, but will keep an eye on it

**Comment by fassadlr** [comment]
@Aprogiena I'm pretty sure we can close this as #825 fixed the logger factory getting disposed issue. I say if it happens again, lets create a new issue?

**Comment by Aprogiena** [comment]
if you think it is solved then ok, feel free to close it, i don't have that knowledge as i did not investigate the issue, so i can't tell

**Comment by fassadlr** [comment]
Ok great! If this issue happens Again @D3m0nKingx, let us know.

Fixed with #825
